cli/ops/webgpu/buffer.rs

```diff
@@ -0,0 +1,224 @@
+use serde::Deserialize;
+use std::borrow::Cow;
+use std::cell::RefCell;
+use std::rc::Rc;
+
+pub fn init(rt: &mut deno_core::JsRuntime) {
+  super::super::reg_json_sync(
+    rt,
+    "op_webgpu_create_buffer",
+    op_webgpu_create_buffer,
+  );
+  super::super::reg_json_async(
+    rt,
+    "op_webgpu_buffer_get_map_async",
+    op_webgpu_buffer_get_map_async,
+  );
+  super::super::reg_json_sync(
+    rt,
+    "op_webgpu_buffer_get_mapped_range",
+    op_webgpu_buffer_get_mapped_range,
+  );
+  super::super::reg_json_sync(
+    rt,
+    "op_webgpu_buffer_unmap",
+    op_webgpu_buffer_unmap,
+  );
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct CreateBufferArgs {
+  instance_rid: u32,
+  device_rid: u32,
+  label: Option<String>,
+  size: u64,
+  usage: u32,
+  mapped_at_creation: Option<bool>,
+}
+
+pub fn op_webgpu_create_buffer(
+  state: &mut OpState,
+  args: Value,
+  _zero_copy: &mut [ZeroCopyBuf],
+) -> Result<Value, AnyError> {
+  let args: CreateBufferArgs = serde_json::from_value(args)?;
+
+  let instance = state
+    .resource_table
+    .get_mut::<super::WgcInstance>(args.instance_rid)
+    .ok_or_else(bad_resource_id)?;
+  let device = state
+    .resource_table
+    .get_mut::<wgc::id::DeviceId>(args.device_rid)
+    .ok_or_else(bad_resource_id)?;
+
+  let descriptor = wgc::resource::BufferDescriptor {
+    label: args.label.map(|label| Cow::Owned(label)),
+    size: args.size,
+    usage: wgt::BufferUsage::from_bits(args.usage).unwrap(),
+    mapped_at_creation: args.mapped_at_creation.unwrap_or(false),
+  };
+  let buffer = wgc::gfx_select!(*device => instance.device_create_buffer(
+    *device,
+    &descriptor,
+    std::marker::PhantomData
+  ))?;
+
+  let rid = state.resource_table.add("webGPUBuffer", Box::new(buffer));
+
+  Ok(json!({
+    "rid": rid,
+  }))
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct BufferGetMapAsyncArgs {
+  instance_rid: u32,
+  buffer_rid: u32,
+  mode: u32,
+  offset: u64,
+  size: u64,
+}
+
+pub async fn op_webgpu_buffer_get_map_async(
+  state: Rc<RefCell<OpState>>,
+  args: Value,
+  _bufs: BufVec,
+) -> Result<Value, AnyError> {
+  let args: BufferGetMapAsyncArgs = serde_json::from_value(args)?;
+
+  let mut state = state.borrow_mut();
+  let instance = state
+    .resource_table
+    .get_mut::<super::WgcInstance>(args.instance_rid)
+    .ok_or_else(bad_resource_id)?;
+  let buffer = state
+    .resource_table
+    .get_mut::<wgc::id::BufferId>(args.buffer_rid)
+    .ok_or_else(bad_resource_id)?;
+
+  let (sender, receiver) = oneshot::channel::<Result<(), AnyError>>();
+
+  let boxed_sender = Box::new(sender);
+  let sender_ptr = Box::into_raw(boxed_sender);
+  let sender_ptr = unsafe {
+    std::mem::transmute::<*mut oneshot::Sender<Result<(), AnyError>>, *mut u8>(
+      sender_ptr,
+    )
+  };
+
+  extern "C" fn buffer_map_future_wrapper(
+    status: wgc::resource::BufferMapAsyncStatus,
+    user_data: *mut u8,
+  ) {
```

Comment:
Not sure if this is correct

Reply:
Doesnt seem to be; just tried and it hangs